[Issue #280] Add TerraformImportCommand and TerraformImportPlugin

duckpuppy opened the pull request:
Partial fix for #280 

This adds a basic TerraformImportPlugin and TerraformImportCommand as an initial cut at implementing the full features of Issue #280.

kmanning commented on src/TerraformImportPlugin.groovy:
```diff
@@ -0,0 +1,49 @@
+        BuildWithParametersPlugin.withStringParameter([
+            name: "IMPORT_TARGET_PATH",
+            description: "The path in the Terraform state to import the spcified resource to."
+        ])
```
Would this work when there is more than one environment?  I thought BuildWithParametersPlugin prompts the user at the start of the pipeline.  If I have QA/UAT/Prod environments in my pipeline, won't it be almost guaranteed that the resource ARNs will be different per environment?  Wouldn't the user need to be prompted on a per-environment basis?

duckpuppy replied:
That is a problem, I think. I'll see what can be done to mitigate that...

kmanning replied:
It looks like you addressed this problem by providing a third `IMPORT_ENVIRONMENT` parameter.  Is that correct?

duckpuppy replied:
That's correct - it's the simplest MVP I came up with. It's not my favorite, but it does work to fix the problem.

kmanning commented:
I had Issue #280 pegged for a v6.0 release, because I assumed it would take a big refactoring of the underlying behavior.  <-- if so, merging this PR should likely wait until v5.16 is closed out.  Perhaps this PR should even be broken in a smaller PR's (depending on the changes needed to support).

Unless you can figure out a clever/small way to implement this behavior without a huge underlying change.  If so, it can go into either v5.16 or v5.17.

duckpuppy commented:
I'm still working on a way to either select from available environments at the start of the build, or present parameters to the user before each run (as well as weighing which of those approaches is the most useful in this case).

kmanning commented:
Pushing to the next release

duckpuppy commented:
This PR is now ready to merge as an MVP - the environment to run the import in is now specified as an additional input parameter.

kmanning commented on docs/TerraformImportPlugin.md:
```diff
@@ -0,0 +1,29 @@
+
+* `IMPORT_RESOURCE`: This is the resource identifier to import into state.
+* `IMPORT_TARGET_PATH`: This is the Terraform state path into which the
+  resource should be imported.
```
There should be a third `IMPORT_ENVIRONMENT` here right?

duckpuppy replied:
Missed that... added.

duckpuppy commented:
That's correct - it's the simplest MVP I came up with. It's not my
favorite, but it does work to fix the problem.

On Tue, Jun 1, 2021 at 9:28 AM kmanning ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/TerraformImportPlugin.groovy
> <https://github.com/manheim/terraform-pipeline/pull/355#discussion_r643103131>
> :
>
> > @@ -0,0 +1,49 @@
> +import static TerraformEnvironmentStage.PLAN_COMMAND
> +
> +class TerraformImportPlugin implements TerraformEnvironmentStagePlugin, TerraformImportCommandPlugin {
> +
> +    public static void init() {
> +        TerraformImportPlugin plugin = new TerraformImportPlugin()
> +
> +        BuildWithParametersPlugin.withStringParameter([
> +            name: "IMPORT_RESOURCE",
> +            description: "Run `terraform import` on the resource specified prior to planning and applying."
> +        ])
> +        BuildWithParametersPlugin.withStringParameter([
> +            name: "IMPORT_TARGET_PATH",
> +            description: "The path in the Terraform state to import the spcified resource to."
> +        ])
>
> It looks like you addressed this problem by providing a third
> IMPORT_ENVIRONMENT parameter. Is that correct?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/manheim/terraform-pipeline/pull/355#discussion_r643103131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEWNPFRNO4YQZ7OKR6BM3TQTOBNANCNFSM4ZZKGRBQ>
> .
>


-- 
SELECT * FROM users WHERE clue > 0

kmanning commented:
There no coverage on `public void apply(TerraformImportCommand command)` and `public Closure runTerraformImportCommand(String environment)` of `TerraformImportPlugin`.

Would it make sense to exercise those methods?  

It feels like for `public void apply(TerraformImportCommand command)`
1. When resource exists, but target missing, import command does not get a target or resource
2. When resource is missing, but target exists, import command does not get a target or resource
2. When resource exists, and target exists, import command gets a target and a resource

It feels like for `runTerraformImportCommand`
1. When a resource exists, a shell command importing the resource should be called on the WorkflowScript
2. When a resource does not exist, no shell commands are called on the WorkflowScript
3. Possibly most importantly, the innerClosure is called

duckpuppy commented:
> There's no coverage on `public void apply(TerraformImportCommand command)` and `public Closure runTerraformImportCommand(String environment)` of `TerraformImportPlugin`.
> 
> Would it make sense to exercise those methods?
> 
> It feels like for `public void apply(TerraformImportCommand command)`
> 
> 1. When resource exists, but target missing, import command does not get a target or resource
> 2. When resource is missing, but target exists, import command does not get a target or resource
> 3. When resource exists, and target exists, import command gets a target and a resource

Added tests to cover these situations.

> It feels like for `runTerraformImportCommand`
> 
> 1. When a resource exists, a shell command importing the resource should be called on the WorkflowScript
> 2. When a resource does not exist, no shell commands are called on the WorkflowScript
> 3. Possibly most importantly, the innerClosure is called

Added tests for these as well, and found a potential edge case with the logic, which is now also corrected.
